// @adobe/uix-guest/dist/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  GuestServer: () => GuestServer,
  GuestUI: () => GuestUI,
  PrimaryGuest: () => GuestServer,
  UIGuest: () => GuestUI,
  attach: () => attach,
  createGuest: () => createGuest,
  register: () => register
});
module.exports = __toCommonJS(src_exports);

// src/guest.ts
var import_uix_core2 = require("@adobe/uix-core");

// src/debug-guest.ts
var import_uix_core = require("@adobe/uix-core");
function debugGuest(guest) {
  return (0, import_uix_core.debugEmitter)(guest, {
    theme: "yellow medium",
    type: "Guest"
  }).listen("beforeconnect", (log, { detail: { guest: guest2 } }) => {
    log.info(guest2);
  }).listen("connecting", (log, { detail: { connection } }) => {
    log.info(connection);
  }).listen("connected", (log, { detail: { guest: guest2 } }) => {
    log.info(guest2);
  }).listen("error", (log, { detail: { error, guest: guest2 } }) => {
    log.error(
      "\u274C Failed to connect! %s",
      error.message,
      guest2,
      error
    );
  });
}

// src/guest.ts
var SharedContext = class {
  constructor(values) {
    this.reset(values);
  }
  reset(values) {
    this._map = new Map(Object.entries(values));
  }
  /**
   * @public
   * Retrieve a copy of a value from the {@link @adobe/uix-host#HostConfig.sharedContext} object. *Note that this is not a reference to any actual objects from the parent. If the parent updates an "inner object" inside the SharedContext, that change will not be reflected in the Guest!*
   */
  get(key) {
    return this._map.get(key);
  }
};
var Guest2 = class extends import_uix_core2.Emitter {
  /**
   * @param config - Initializer for guest object, including ID.
   */
  constructor(config) {
    super(config.id);
    this.logger = import_uix_core2.quietConsole;
    /**
     * Proxy object for calling methods on the host.
     *
     * @remarks Any APIs exposed to the extension via {@link @adobe/uix-host#Port.provide}
     * can be called on this object. Because these methods are called with RPC,
     * they are all asynchronous, The return types of all Host methods will be
     * Promises which resolve to the value the Host method returns.
     * @public
     */
    this.host = (0, import_uix_core2.makeNamespaceProxy)(
      async (address) => {
        await this.hostConnectionPromise;
        try {
          const result = await (0, import_uix_core2.timeoutPromise)(
            () => `Calling ${(0, import_uix_core2.formatHostMethodAddress)(address)}`,
            this.invokeAwaiter(
              this.hostConnection.getRemoteApi().invokeHostMethod,
              address
            ),
            1e4
          );
          return result;
        } catch (e) {
          const error = e instanceof Error ? e : new Error(e);
          this.logger.error(error);
          throw error;
        }
      }
    );
    this.timeout = 2e4;
    if (typeof config.timeout === "number") {
      this.timeout = config.timeout;
    }
    if (config.debug) {
      this.logger = debugGuest(this);
    }
    this.addEventListener("contextchange", (event) => {
      this.sharedContext = new SharedContext(event.detail.context);
    });
  }
  /**
   * @internal
   */
  async invokeChecker(invoker, address) {
    try {
      const res = await invoker(address);
      return new Promise((resolve) => resolve(res));
    } catch (e) {
      await new Promise((resolve) => setTimeout(resolve, 500));
      return this.invokeChecker(invoker, address);
    }
  }
  /**
   * @internal
   */
  async invokeAwaiter(invoker, address) {
    const final = setTimeout(() => {
      return new Promise(
        (resolve, reject) => reject(`${address} doesn't exist`)
      );
    }, 2e4);
    const res = await this.invokeChecker(invoker, address);
    return new Promise((resolve) => {
      clearTimeout(final);
      return resolve(res);
    }).catch((e) => {
      clearTimeout(final);
      return e;
    });
  }
  /** @internal */
  getLocalMethods() {
    return {
      emit: (...args) => {
        this.logger.log(`Event "${args[0]}" emitted from host`);
        this.emit(...args);
      }
    };
  }
  /**
   * Accept a connection from the Host.
   * @returns A Promise that resolves when the Host has established a connection.
   * @deprecated It is preferable to use {@link register} for primary frames,
   * and {@link attach} for UI frames and other secondary frames, than to
   * instantiate a Guest and then call `.connect()` on it. The latter style
   * returns an object that cannot be used until it is connected, and therefore
   * risks errors.
   * @public
   */
  async connect() {
    return this._connect();
  }
  /**
   * @internal
   */
  async _connect() {
    this.emit("beforeconnect", { guest: this });
    try {
      const hostConnectionPromise = (0, import_uix_core2.connectParentWindow)(
        {
          targetOrigin: "*",
          timeout: this.timeout,
          logger: this.logger
        },
        this.getLocalMethods()
      );
      this.hostConnectionPromise = hostConnectionPromise;
      this.hostConnection = await this.hostConnectionPromise;
      this.emit("connected", { guest: this });
    } catch (e) {
      this.emit("error", { guest: this, error: e });
      this.logger.error("Connection failed!", e);
      throw e;
    }
    try {
      this.sharedContext = new SharedContext(
        await this.hostConnection.getRemoteApi().getSharedContext()
      );
    } catch (e) {
      this.emit("error", { guest: this, error: e });
      this.logger.error("getSharedContext failed!", e);
      throw e;
    }
    try {
      this.configuration = await this.hostConnection.getRemoteApi().getConfiguration();
    } catch (e) {
      this.emit("error", { guest: this, error: e });
      this.logger.error("getConfiguration failed!", e);
      throw e;
    }
    if (window.parent && window.parent !== window) {
      window.parent.postMessage({ type: "guest-ready", guestId: this.id }, "*");
    }
  }
};

// src/guest-ui.ts
var GuestUI = class extends Guest2 {
  /**
   * {@inheritDoc Guest."constructor"}
   */
  constructor(config) {
    super(config);
    this.addEventListener("connected", () => {
      const resizeObserver = new ResizeObserver((entries) => {
        const doc = entries.find((entry) => entry.target === document.body);
        const borderBoxSize = doc.borderBoxSize.length ? doc.borderBoxSize[0] : doc.borderBoxSize;
        this.hostConnection.getRemoteApi().onIframeResize({
          height: borderBoxSize.blockSize + this.calculateChildrenMargin(doc.target.querySelectorAll("*")),
          width: borderBoxSize.inlineSize
        });
      });
      resizeObserver.observe(document.body);
    });
    this.logger.log("Will add resize observer on connect");
  }
  /**
   * @internal
   */
  calculateChildrenMargin(elems) {
    let margin = 0;
    for (let i = 0; i < elems.length; i++) {
      const style = elems[i].currentStyle || window.getComputedStyle(elems[i]);
      const marginTop = parseInt(style.marginTop);
      const marginBottom = parseInt(style.marginBottom);
      if (marginTop > 0) {
        margin = margin + marginTop;
      }
      if (marginBottom > 0) {
        margin = margin + marginBottom;
      }
    }
    return margin;
  }
};

// src/guest-server.ts
var GuestServer = class extends Guest2 {
  getLocalMethods() {
    return {
      ...super.getLocalMethods(),
      apis: this.localMethods,
      metadata: this.metadata
    };
  }
  /**
   * Pass an interface of methods which Host may call as callbacks.
   *
   * @remarks It is preferable to use {@link register} to obtain a guest object
   * and register local methods in one step. The returned guest object will be
   * pre-registered and connected.
   * @public
   */
  async register(implementedMethods, metadata) {
    this.localMethods = implementedMethods;
    this.metadata = {
      ...metadata,
      extensionId: this.id
    };
    return this._connect();
  }
};

// src/index.ts
function createGuest(config) {
  const guest = new GuestServer(config);
  return guest;
}
async function attach(config) {
  const guest = new GuestUI(config);
  await guest._connect();
  return guest;
}
async function register(config) {
  const guest = new GuestServer(config);
  guest.register(config.methods, config.metadata);
  return guest;
}
